docs/01-tooling/01-automatic-file-bundling.md

```diff
@@ -1,6 +1,188 @@
+system, so files have to expose and consume other files by setting properties
+and objects on global variables, and getting properties and variables from
+global variables.
+
+This has caused issues in the development workflows related to the project's
+JavaScript sources, and the use of JavaScript libraries from the OSS ecosystem.
+
+#### Application JavaScript sources
+
+The order in which this files needs to be loaded so that all dependencies are
+set properly then needs to be [manually specified in
+`extension.json`][scripts-order] by file name.  As such, there are two sources
+of truth:
+
+* The configuration that specifies some order of files,
+* And the source code that implicitly uses files in some order.
+
+If you are doing anything mildly complex, you will end up with a big list of
+JavaScript files that depend on each other. Having to manually understand that
+dependency graph based on the list of dependencies on a JSON file and how the
+files use, define, or re-define global variables from or for other files, can be
+quite of a headache.
+
+Changes to the source code, moving lines in the same file, refactoring code to
+other and new files, adding new code that uses other files, removing code, ...
+
+All of these become really hard really quickly. In the end, organizing your
+code, trying to keep it maintainable, and understanding how it is structured
+have a very big cost on the project's quality, development and maintenance.
+
+Figuring out if the files are specified in order properly in the configuration
+after and while you make changes has a high potential for obscure runtime bugs
+by forgetting or not seeing implicit dependencies in the manual order of files.
+
+#### Libraries
+
+In order to consume libraries for the front-end code, right now we rely on
+manually (or by script) pulling down npm dependencies or files from a website
+and including them in the repository. This is hard to verify, keep up to date,
+and cumbersome. If necessary, it should be possible to specify dependencies with
+the versions and have them be automatically included in the assets we serve from
+a trustworthy source. It should be easy to check if the libraries are outdated,
+and update them.
+
 ### Requirements
+
+For developing JavaScript files for the front-end:
+
+* There must be a single source of truth for the dependency resolution
+* Files should be authored in a standard module system instead of relying in the
+  global namespace
+* Moving code and files should be low cost and not trigger runtime errors in
+  production if the dependencies are not properly specified
+* If possible, it should be easier to tap into npm libraries for our front-end
+  needs, and check for updates
+
 ### Solutions
+
+We considered our options, and after discussion among the engineers, we decided
+that:
+
+* We would use a node based file bundler ([webpack][]), in order to:
+  * automatically bundle our JS sources with a single source of truth (the
```

Comment:
The "single source of truth" statement needs a little more detail here so that it can be linked back to the section above. How about "a single source of truth for intra-module dependencies and asset load order"?

Reply:
Sounds good, changed to:

```
  * automatically bundle our JS sources with a single source of truth for
    intra-module dependencies and asset load order (the source code itself via
    import/export statements)
```